File: redshirt/interaction.py
"""Local JSON-lines decisions and a small async client. No environment rules.

The existing runner owns execution, timeouts, checks, evidence and replay. This
transport only exposes its audited selector request and accepts one choice.
"""
import asyncio
import json
import secrets
import signal
import sys

from .evidence import encoded
from .runner import Limits, Stop, run

MAX_MESSAGE = 32768
MAX_REPLY = 4096


def choice_tool(candidates):
    """Provider-neutral function definition; the controller keeps decision IDs."""
    return {"name": "choose_action", "description": "Choose one currently offered action.",
            "parameters": {"type": "object", "properties": {
                "action_id": {"type": "string", "enum": list(candidates)}},
                "required": ["action_id"], "additionalProperties": False}}


def _decode(raw, limit):
    def nonfinite(_):
        raise ValueError("nonfinite")

    def unique(pairs):
        result = {}
        for key, value in pairs:
            if key in result:
                raise ValueError("duplicate_key")
            result[key] = value
        return result
    if not raw or len(raw) > limit or not raw.endswith(b"\n"):
        raise ValueError("message_size_or_framing")
    result = json.loads(raw, object_pairs_hook=unique, parse_constant=nonfinite)
    if not isinstance(result, dict):
        raise ValueError("message_shape")
    return result


async def _send(writer, message):
    raw = encoded(message) + b"\n"
    if len(raw) > MAX_MESSAGE:
        raise ValueError("message_size")
    writer.write(raw)
    await writer.drain()


class Interactive:
    """One response per observation; EOF or invalid replies stop, never retry."""
    name = "local-jsonl"
    evidence_limit = 2048

    def __init__(self, reader, writer):
        self.reader, self.writer = reader, writer
        self._busy = False
        self._receipts = []

    async def select(self, request):
        if self._busy:
            raise Stop("provider_busy")
        self._busy = True
        decision = secrets.token_hex(16)
        receipt = {"decision_id": decision, "status": "interrupted"}
        try:
            await _send(self.writer, {**request, "type": "observation", "decision_id": decision,
                                      "tool": choice_tool(request["candidates"])})
            try:
                raw = await self.reader.readline()
                if not raw:
                    raise Stop("client_disconnected")
                reply = _decode(raw, MAX_REPLY)
            except (ValueError, UnicodeError):
                raise Stop("invalid_api_reply") from None
            if (set(reply) != {"decision_id", "action_id"}
                    or not isinstance(reply["decision_id"], str)
                    or not isinstance(reply["action_id"], str)):
                raise Stop("invalid_api_reply")
            if reply["decision_id"] != decision:
                raise Stop("stale_decision")
            if reply["action_id"] not in request["candidates"]:
                raise Stop("unknown_candidate")
            receipt.update(status="selected", action_id=reply["action_id"])
            return reply["action_id"]
        except Stop as exc:
            receipt["status"] = str(exc)
            raise
        except (BrokenPipeError, ConnectionError):
            receipt["status"] = "client_disconnected"
            raise Stop("client_disconnected") from None
        finally:
            self._receipts.append(receipt)
            self._busy = False

    def take_evidence(self):
        receipts, self._receipts = self._receipts, []
        return receipts


def completion(report):
    """No evaluator internals, private operations or artifact paths on the wire."""
    return {"version": 1, "type": "done", "stop": report["stop"],
            "requests": report["requests"], "attempted_inputs": report["attempted_inputs"],
            "verified": report["final"]["ok"], "cleanup": report["cleanup"],
            "replayable": report["replayable"]}


async def serve_stdio(adapter, output, *, limits=Limits(captures=0), cancel=None):
    """Run one bounded episode over local pipes. Stdout is exclusively JSONL.

    POSIX pipe transport; no port, remote browser or model credential is opened.
    Decision time continues to advance in real-time environments.
    """
    loop = asyncio.get_running_loop()
    reader = asyncio.StreamReader(limit=MAX_REPLY)
    incoming, _ = await loop.connect_read_pipe(lambda: asyncio.StreamReaderProtocol(reader), sys.stdin.buffer)
    outgoing = None
    try:
        outgoing, protocol = await loop.connect_write_pipe(
            lambda: asyncio.streams.FlowControlMixin(loop=loop), sys.stdout.buffer)
        writer = asyncio.StreamWriter(outgoing, protocol, None, loop)
        report = await run(adapter, output, provider=Interactive(reader, writer), limits=limits, cancel=cancel)
        try:
            await asyncio.wait_for(_send(writer, completion(report)), 5)
        except (TimeoutError, BrokenPipeError, ConnectionError):
            pass  # The runner has already preserved checks and finalized evidence.
        return report
    finally:
        incoming.close()
        if outgoing:
            outgoing.close()


class InteractionClient:
    """Async client for an operator-selected local adapter process.

    Call observe(), choose from candidates/tool, then await act(action_id).
    The client inserts the decision token. It never accepts model-supplied argv.
    act() returns the next observation or a terminal done envelope.
    """
    def __init__(self, process):
        self.process = process
        self._frame = None
        self._busy = False

    @classmethod
    async def start(cls, *argv, cwd=None, env=None):
        process = await asyncio.create_subprocess_exec(*argv, cwd=cwd, env=env,
            stdin=asyncio.subprocess.PIPE, stdout=asyncio.subprocess.PIPE, limit=MAX_MESSAGE)
        client = cls(process)
        try:
            client._frame = await client._receive()
            return client
        except BaseException:
            await client.close()
            raise

    async def _receive(self):
        raw = await asyncio.wait_for(self.process.stdout.readline(), 35)
        message = _decode(raw, MAX_MESSAGE)
        if message.get("version") != 1 or message.get("type") not in ("observation", "done"):
            raise ValueError("invalid_api_message")
        return message

    def observe(self):
        if self._frame is None:
            raise ValueError("no_current_observation")
        return json.loads(encoded(self._frame))

    async def act(self, action_id):
        if self._busy:
            raise ValueError("client_busy")
        frame = self.observe()
        if frame["type"] != "observation":
            raise ValueError("session_finished")
        if not isinstance(action_id, str) or action_id not in frame["candidates"]:
            raise ValueError("unknown_candidate")
        self._busy, self._frame = True, None
        try:
            await _send(self.process.stdin, {"decision_id": frame["decision_id"], "action_id": action_id})
            self._frame = await self._receive()
            return self.observe()
        finally:
            self._busy = False

    async def close(self):
        if self.process.returncode is None:
            try:
                self.process.send_signal(signal.SIGINT)
            except ProcessLookupError:
                pass
            try:
                await asyncio.wait_for(self.process.wait(), 30)
            except TimeoutError:
                self.process.kill()
                await self.process.wait()
        if self.process.stdin:
            self.process.stdin.close()

    async def __aenter__(self):
        return self

    async def __aexit__(self, *_):
        await self.close()

File: src/interaction.rs
//! The existing local decision protocol. No environment authority or game rules.
use crate::{Provider, Result, Stop, evidence, require};
use async_trait::async_trait;
use serde_json::{Value, json};
use tokio::io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt, BufReader};

pub const MAX_MESSAGE: usize = 32768;
pub const MAX_REPLY: usize = 4096;

pub fn choice_tool(candidates: &serde_json::Map<String, Value>) -> Value {
    json!({"name":"choose_action","description":"Choose one currently offered action.",
        "parameters":{"type":"object","properties":{"action_id":{
            "type":"string","enum":candidates.keys().collect::<Vec<_>>() }},
            "required":["action_id"],"additionalProperties":false}})
}

pub fn completion(report: &Value) -> Value {
    json!({"version":1,"type":"done","stop":report["stop"],
        "requests":report["requests"],"attempted_inputs":report["attempted_inputs"],
        "verified":report["final"]["ok"],"cleanup":report["cleanup"],
        "replayable":report["replayable"]})
}

pub struct Interactive<R, W> {
    reader: BufReader<R>,
    writer: W,
    incoming: Vec<u8>,
    outgoing: Vec<u8>,
    written: usize,
    receipt: Option<Value>,
}

impl<R: AsyncRead + Unpin, W: AsyncWrite + Unpin> Interactive<R, W> {
    pub fn new(reader: R, writer: W) -> Self {
        Self {
            reader: BufReader::new(reader),
            writer,
            incoming: vec![],
            outgoing: vec![],
            written: 0,
            receipt: None,
        }
    }

    // Keep progress outside the future: cancellation must not splice JSON frames.
    async fn flush_pending(&mut self) -> Result<()> {
        while self.written < self.outgoing.len() {
            let n = self
                .writer
                .write(&self.outgoing[self.written..])
                .await
                .map_err(|_| Stop::from("client_disconnected"))?;
            require(n != 0, "client_disconnected")?;
            self.written += n;
        }
        self.writer
            .flush()
            .await
            .map_err(|_| Stop::from("client_disconnected"))?;
        self.outgoing.clear();
        self.written = 0;
        Ok(())
    }

    async fn send(&mut self, message: &Value) -> Result<()> {
        self.flush_pending().await?;
        let mut raw = evidence::encoded(message)?;
        raw.push(b'\n');
        require(raw.len() <= MAX_MESSAGE, "api_message_size")?;
        self.outgoing = raw;
        self.flush_pending().await
    }

    async fn receive(&mut self) -> Result<Value> {
        loop {
            let chunk = self
                .reader
                .fill_buf()
                .await
                .map_err(|_| Stop::from("client_disconnected"))?;
            if chunk.is_empty() {
                return Err(if self.incoming.is_empty() {
                    "client_disconnected"
                } else {
                    "invalid_api_reply"
                }
                .into());
            }
            let newline = chunk.iter().position(|b| *b == b'\n');
            let n = newline.map_or(chunk.len(), |i| i + 1);
            require(self.incoming.len() + n <= MAX_REPLY, "invalid_api_reply")?;
            self.incoming.extend_from_slice(&chunk[..n]);
            self.reader.consume(n);
            if newline.is_some() {
                return evidence::decode(&std::mem::take(&mut self.incoming))
                    .map_err(|_| "invalid_api_reply".into());
            }
        }
    }

    pub async fn finish(&mut self, report: &Value) -> Result<()> {
        self.send(&completion(report)).await
    }
}

#[async_trait]
impl<R: AsyncRead + Unpin + Send, W: AsyncWrite + Unpin + Send> Provider for Interactive<R, W> {
    fn name(&self) -> &str {
        "local-jsonl"
    }
    fn evidence_limit(&self) -> usize {
        2048
    }
    async fn select(&mut self, request: Value) -> Result<String> {
        require(self.receipt.is_none(), "provider_busy")?;
        let mut random = [0; 16];
        getrandom::fill(&mut random).map_err(|_| Stop::from("decision_entropy"))?;
        let decision: String = random.iter().map(|b| format!("{b:02x}")).collect();
        // Installed before awaiting, so a dropped future retains its receipt.
        self.receipt = Some(json!({"decision_id":decision,"status":"interrupted"}));
        let result: Result<String> = async {
            let candidates = request["candidates"]
                .as_object()
                .ok_or_else(|| Stop::from("invalid_request"))?;
            let mut message = request
                .as_object()
                .ok_or_else(|| Stop::from("invalid_request"))?
                .clone();
            message.insert("type".into(), "observation".into());
            message.insert("decision_id".into(), decision.clone().into());
            message.insert("tool".into(), choice_tool(candidates));
            self.send(&Value::Object(message)).await?;
            let reply = self.receive().await?;
            let obj = reply
                .as_object()
                .ok_or_else(|| Stop::from("invalid_api_reply"))?;
            require(
                obj.len() == 2
                    && reply["decision_id"].is_string()
                    && reply["action_id"].is_string(),
                "invalid_api_reply",
            )?;
            require(reply["decision_id"] == decision, "stale_decision")?;
            let action = reply["action_id"].as_str().expect("validated string");
            require(candidates.contains_key(action), "unknown_candidate")?;
            Ok(action.into())
        }
        .await;
        let receipt = self.receipt.as_mut().expect("pending receipt");
        match &result {
            Ok(action) => {
                receipt["status"] = "selected".into();
                receipt["action_id"] = action.clone().into();
            }
            Err(error) => receipt["status"] = error.to_string().into(),
        }
        result
    }
    fn take_evidence(&mut self) -> Vec<Value> {
        self.receipt.take().into_iter().collect()
    }
}

/// Async pipe descriptors avoid an uncancellable blocking stdin read at shutdown.
#[cfg(unix)]
pub fn stdio()
-> Result<Interactive<tokio::net::unix::pipe::Receiver, tokio::net::unix::pipe::Sender>> {
    use std::os::fd::AsFd;
    use tokio::net::unix::pipe::{Receiver, Sender};
    let incoming = std::io::stdin()
        .as_fd()
        .try_clone_to_owned()
        .map_err(|_| Stop::from("stdio_requires_pipes"))?;
    let outgoing = std::io::stdout()
        .as_fd()
        .try_clone_to_owned()
        .map_err(|_| Stop::from("stdio_requires_pipes"))?;
    Ok(Interactive::new(
        Receiver::from_owned_fd(incoming).map_err(|_| Stop::from("stdio_requires_pipes"))?,
        Sender::from_owned_fd(outgoing).map_err(|_| Stop::from("stdio_requires_pipes"))?,
    ))
}
